collector/agent_poller.go

```diff
@@ -69,80 +61,120 @@ type AgentState struct {
 	executorAppNames map[string]string
 }
 
+type Agent struct {
+	AgentIP        string
+	IPCommand      string
+	Port           int
+	PollPeriod     int
+	Topic          string
+	AgentStateChan chan<- *AgentState
+}
+
+func NewAgent(
+	ipCommand string,
+	port int,
+	pollPeriod int,
+	topic string) (Agent, error) {
+	a := Agent{}
+	if len(ipCommand) < 0 {
+		return a, errors.New("Must pass ipAddress to NewAgent()")
+	}
+	if port < 1024 {
+		return a, errors.New("Must pass port to NewAgent()")
+	}
+	if pollPeriod == 0 {
+		return a, errors.New("Must pass pollPeriod to NewAgent()")
+	}
+	if len(topic) < 1 {
+		return a, errors.New("Must pass topic to NewAgent()")
+	}
+
+	a.IPCommand = ipCommand
+	a.Port = port
+	a.PollPeriod = pollPeriod
+	a.Topic = topic
+	a.AgentStateChan = make(chan *AgentState)
+
+	return a, nil
+}
+
 // Runs an Agent Poller which periodically produces data retrieved from a local Mesos Agent.
 // This function should be run as a gofunc.
-func RunAgentPoller(recordsChan chan<- *AvroDatum, agentStateChan chan<- *AgentState, stats chan<- StatsEvent) {
+func (a *Agent) Run(recordsChan chan<- *AvroDatum, stats chan<- StatsEvent) {
 	// fetch agent ip once. per DC/OS docs, changing a node IP is unsupported
-	var agentIp string = ""
+	a.AgentIP = ""
 	if len(*agentTestStateFileFlag) == 0 ||
 		len(*agentTestSystemFileFlag) == 0 ||
 		len(*agentTestContainersFileFlag) == 0 {
 		// only get the ip if actually needed
-		agentIp = getAgentIp(stats)
+		//TODO needs err
+		a.getIP(stats)
 	}
 
 	// do one poll immediately upon starting, to ensure that agent metadata is populated early:
-	pollAgent(agentIp, recordsChan, agentStateChan, stats)
-	ticker := time.NewTicker(time.Second * time.Duration(*agentPollPeriodFlag))
+	a.pollAgent(recordsChan, stats)
+	ticker := time.NewTicker(time.Second * time.Duration(a.PollPeriod))
```

Comment:
rather then passing integer `a.PollPeriod`, can we pass a duration string and read is with `ParseDuration`
https://golang.org/pkg/time/#ParseDuration
you can take a look at this example:
https://blog.golang.org/context Look for this line `time.ParseDuration(req.FormValue("timeout"))`

this will give us much more flexibility